core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationMonitor.scala

```diff
@@ -55,236 +57,392 @@ object WorkspaceMigrationMonitor {
-        sourceBucketName,
-        List(BucketGetOption.userProject(destGoogleProject.value))
-      )
+      migration <- findMigration { m =>
+        m.tmpBucketTransferred.isDefined && m.tmpBucketDeleted.isEmpty
+      }
 
-      sourceBucket <- IO(sourceBucketOpt.getOrElse(throw noWorkspaceBucketError(sourceBucketName)))
+      (googleProjectId, tmpBucketName) <- MigrateAction.liftIO(IO {
+        (
+          migration.newGoogleProjectId.getOrElse(throw noGoogleProjectError(migration)),
+          migration.tmpBucketName.getOrElse(throw noTmpBucketError(migration))
+        )
+      })
 
-      // todo: CA-1637 do we need to transfer the storage logs for this workspace? the logs are prefixed
-      // with the ws bucket name, so we COULD do it, but do we HAVE to? it's a csv with the bucket
-      // and the storage_byte_hours in it that is kept for 180 days
-      _ <- googleStorageService.insertBucket(
-        googleProject = destGoogleProject,
-        bucketName = destBucketName,
-        labels = Option(sourceBucket.getLabels).map(_.toMap).getOrElse(Map.empty),
-        bucketPolicyOnlyEnabled = true,
-        logBucket = Option(GcsBucketName(GoogleServicesDAO.getStorageLogsBucketName(GoogleProjectId(destGoogleProject.value)))),
-        location = Option(sourceBucket.getLocation)
-      ).compile.drain
+      storageService <- MigrateAction.asks(_.storageService)
+      _ <- MigrateAction.liftIO {
+        for {
+          successOpt <- storageService.deleteBucket(
+            GoogleProject(googleProjectId.value),
+            tmpBucketName,
+            isRecursive = true
+          ).compile.last
+
+          _ <- IO.raiseUnless(successOpt.contains(true)) {
+            noTmpBucketError(migration)
+          }
+        } yield ()
+      }
 
+      deleted <- timestampNow
+      _ <- inTransaction { _ =>
+        workspaceMigrations
+          .filter(_.id === migration.id)
+          .map(_.tmpBucketDeleted)
+          .update(deleted.some)
+          .ignore
+      }
     } yield ()
 
 
-  final def startStorageTransferJobToTmpBucket(migration: WorkspaceMigration,
-                                               workspace: Workspace,
-                                               projectToBill: GoogleProject,
-                                               sts: GoogleStorageTransferService[IO]
-                                              ): IO[(TransferJob, WriteAction[Unit])] =
-    IO.raiseWhen(migration.tmpBucketName.isEmpty)(noTmpBucketError(migration)) *>
-      startBucketStorageTransferJob(
-        migration,
-        GcsBucketName(workspace.bucketName),
-        migration.tmpBucketName.get,
-        projectToBill,
-        sts
-      )
+  final def findMigration(predicate: WorkspaceMigrationHistory => Rep[Boolean])
+  : MigrateAction[WorkspaceMigration] =
+    inTransaction { _ =>
+      workspaceMigrations
+        .filter(row => row.finished.isEmpty && predicate(row))
+        .sortBy(_.updated)
+        .take(1)
```

Comment:
Are there any protections to prevent us from creating multiple migration rows for the same workspace?  Uniqueness constraints or something in queries to raise a fuss if we somehow have multiples?

Reply:
A row in the database represents an attempt at migrating a workspace. I've configured the table such that you can't have more than one active (ie non-finished) rows at a time. You CAN have as many finished rows as you like. If you try to create concurrent migration attempts of the same workspace you'll get an error.

Reply:
https://github.com/broadinstitute/rawls/blob/b7252ec2d1826e0d28bc5378d2632c3d00315740/core/src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20211115_v1_workspace_migration_history.xml#L56-L81

Reply:
https://github.com/broadinstitute/rawls/blob/b7252ec2d1826e0d28bc5378d2632c3d00315740/core/src/test/scala/org/broadinstitute/dsde/rawls/monitor/WorkspaceMigrationMonitorSpec.scala#L99-L105